File: not-required/express-backend/src/routes/logs.ts
import express, { Request, Response, Router } from 'express';
import { PrismaClient } from '@prisma/client';
import Joi from 'joi';

const router: Router = express.Router();

// ─── GET /api/logs ────────────────────────────────────────────────────────────
// Returns paginated call logs for the authenticated tenant.
// Query params: agentId, from, to (ISO strings), page, limit
router.get('/', async (req: Request, res: Response) => {
  try {
    const prisma: PrismaClient = req.app.get('prisma');
    const tenantId = req.tenantId;

    const page  = Math.max(1, parseInt(req.query.page  as string) || 1);
    const limit = Math.min(200, Math.max(1, parseInt(req.query.limit as string) || 50));
    const skip  = (page - 1) * limit;

    const where: any = { tenantId };

    if (req.query.agentId) {
      where.agentId = req.query.agentId as string;
    }
    if (req.query.from || req.query.to) {
      where.startedAt = {};
      if (req.query.from) where.startedAt.gte = new Date(req.query.from as string);
      if (req.query.to)   where.startedAt.lte = new Date(req.query.to   as string);
    }

    const [logs, total] = await Promise.all([
      prisma.callLog.findMany({
        where,
        skip,
        take: limit,
        orderBy: { startedAt: 'desc' },
        include: {
          agent: {
            select: { id: true, name: true },
          },
        },
      }),
      prisma.callLog.count({ where }),
    ]);

    res.json({ logs, total, page, limit, pages: Math.ceil(total / limit) });
  } catch (error) {
    console.error('Error fetching call logs:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── PATCH /api/logs/:id/rating ───────────────────────────────────────────────
// Set thumbs-up (1) or thumbs-down (-1) rating + optional notes on a log.
const ratingSchema = Joi.object({
  rating: Joi.number().valid(1, -1).required(),
  notes:  Joi.string().max(2000).allow('', null).optional(),
});

router.patch('/:id/rating', async (req: Request, res: Response) => {
  const { error, value } = ratingSchema.validate(req.body);
  if (error) return res.status(400).json({ error: error.details[0].message });

  try {
    const prisma: PrismaClient = req.app.get('prisma');

    // Verify the log belongs to this tenant before updating
    const log = await prisma.callLog.findUnique({ where: { id: req.params.id } });
    if (!log)                        return res.status(404).json({ error: 'Log not found' });
    if (log.tenantId !== req.tenantId) return res.status(403).json({ error: 'Forbidden' });

    const updated = await prisma.callLog.update({
      where: { id: req.params.id },
      data: {
        rating:      value.rating,
        ratingNotes: value.notes ?? null,
      },
    });

    res.json(updated);
  } catch (error) {
    console.error('Error updating call log rating:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ─── POST /api/logs/:id/flag ──────────────────────────────────────────────────
// Mark a log as flagged for retraining.
router.post('/:id/flag', async (req: Request, res: Response) => {
  try {
    const prisma: PrismaClient = req.app.get('prisma');

    const log = await prisma.callLog.findUnique({ where: { id: req.params.id } });
    if (!log)                        return res.status(404).json({ error: 'Log not found' });
    if (log.tenantId !== req.tenantId) return res.status(403).json({ error: 'Forbidden' });

    const updated = await prisma.callLog.update({
      where: { id: req.params.id },
      data: { flaggedForRetraining: true },
    });

    res.json(updated);
  } catch (error) {
    console.error('Error flagging call log:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: python/backend/app/routes/logs.py
"""
/api/logs routes — mirrors Express src/routes/logs.ts
"""
import math
from fastapi import APIRouter, Depends, Query
from fastapi.responses import JSONResponse
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional

from app.database import get_db
from app.auth import AuthContext, get_auth
from app.models import CallLog, Agent

router = APIRouter()


@router.get("/")
async def list_logs(
    page: int = 1,
    limit: int = 50,
    agentId: Optional[str] = None,
    auth: AuthContext = Depends(get_auth),
    db: AsyncSession = Depends(get_db),
):
    page = max(1, page)
    limit = min(200, max(1, limit))
    where = [CallLog.tenantId == auth.tenant_id]
    if agentId:
        where.append(CallLog.agentId == agentId)

    total = (await db.execute(select(func.count(CallLog.id)).where(*where))).scalar() or 0
    q = (
        select(CallLog, Agent.id.label("a_id"), Agent.name.label("a_name"))
        .outerjoin(Agent, Agent.id == CallLog.agentId)
        .where(*where)
        .order_by(CallLog.startedAt.desc())
        .offset((page - 1) * limit)
        .limit(limit)
    )
    result = await db.execute(q)
    rows = result.all()

    logs = []
    for row in rows:
        log = row[0]
        logs.append({
            "id": log.id,
            "tenantId": log.tenantId,
            "agentId": log.agentId,
            "callerPhone": log.callerPhone,
            "startedAt": log.startedAt.isoformat() if log.startedAt else None,
            "endedAt": log.endedAt.isoformat() if log.endedAt else None,
            "durationSeconds": log.durationSeconds,
            "transcript": log.transcript,
            "analysis": log.analysis,
            "rating": log.rating,
            "ratingNotes": log.ratingNotes,
            "flaggedForRetraining": log.flaggedForRetraining,
            "retrained": log.retrained,
            "createdAt": log.createdAt.isoformat() if log.createdAt else None,
            "agent": {"id": row[1], "name": row[2]} if row[1] else None,
        })

    return {"logs": logs, "total": total, "page": page, "limit": limit, "pages": math.ceil(total / limit) if limit else 0}


@router.patch("/{log_id}/rating")
async def rate_log(log_id: str, body: dict, auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    rating = body.get("rating")
    if rating not in (1, -1):
        return JSONResponse({"error": "rating must be 1 or -1"}, status_code=400)

    result = await db.execute(select(CallLog).where(CallLog.id == log_id))
    log = result.scalar_one_or_none()
    if not log:
        return JSONResponse({"error": "Log not found"}, status_code=404)
    if log.tenantId != auth.tenant_id:
        return JSONResponse({"error": "Forbidden"}, status_code=403)

    log.rating = rating
    log.ratingNotes = body.get("notes")
    await db.commit()
    await db.refresh(log)
    return {"id": log.id, "rating": log.rating, "ratingNotes": log.ratingNotes}


@router.post("/{log_id}/flag")
async def flag_log(log_id: str, auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(CallLog).where(CallLog.id == log_id))
    log = result.scalar_one_or_none()
    if not log:
        return JSONResponse({"error": "Log not found"}, status_code=404)
    if log.tenantId != auth.tenant_id:
        return JSONResponse({"error": "Forbidden"}, status_code=403)

    log.flaggedForRetraining = True
    await db.commit()
    return {"id": log.id, "flaggedForRetraining": True}
